### ISSUE TITLE
Restart testing builds immediately after a push to master

### ISSUE THREAD

**Issue by rudymatela** [channable/hoff#183]
We should restart `testing/...` builds immediately after a push to master by listening for the ["push GitHub webhook](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push) to reduce promotion times and CI minutes.  The detailed explanation follows.

------

When a build succeeds :heavy_check_mark:  but `master` has been updated since, the (promotion) push to `master` fails.  This is what we see on the comment timeline:

> Rebased as [532c843](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

... 4 minutes later ...

[Unknown to Hoff `master` is updated.]

... 6 minutes later...

> Rebased as [a348c32](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

... 10 minutes later the build passes :heavy_check_mark: and Hoff merges.

------

Instead of waiting for the build to finish to discover that the `testing/...` branch is stale, Hoff could listen for the ["push" GitHub webkook](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push) and restart the `testing/...` build right away.

------

If we implement the idea above, this is what we _should_ see on the comment timeline:

> Rebased as [532c843](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

... 4 minutes later ...

[`master` is updated.]

> Testing branch `testing/...` is stale due to a push to `master`.  Rebasing again...

> Rebased as [a348c32](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

... 10 minutes later the build passes :heavy_check_mark: and Hoff merges.

------

Without the proposed change, we waited a total of 20 minutes to merge while wasting 6 minutes of CI build time.

With the proposed change, we only wait a total of 14 minutes and waste no CI build time.  The end result is also less confusing, as Hoff would post a message explaining the duplicate rebase.

This should also reduce the chance of an unfortunate sequence of failed pushes to master that currently cause the timeline to look like this in seldom occasions:

> Rebased as [532c843](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

> Rebased as [0123456](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

> Rebased as [abc0123](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

> Rebased as [321dcba](https://github.com/channable/hoff/commit/532c843c7f1f827e50ebc9c466791a81dfec1ab7), waiting for CI …

> [CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/9d117f73-dc76-4ff8-9ebd-751fe990151c?pipeline_id=0b430733-1e3b-4612-bce9-4b032fdb846a) started.

Merged.

**Comment by rudymatela** [comment]
@bertptrs :point_up:

**Comment by tijmendj** [comment]
This sounds really good, but what I'm wondering is what happens when multiple branches all restart CI at the same time. Forgive my ASCII art, but I think this is now the case:

(Capitals are branches, `c` is CI starting, `m` is 'merged')

```
A->c---->m
--B->c---->c---->m
----C->c---->c----c--->m
```
Proposed situation:
```
A->c---->m
--B->---->c---->?
---->C--->c---->?
```

Won't every merge start a a race between all of the other branches that are in the process of being merged? Is that a problem aside from making it pretty much random what gets merged first?

**Comment by rudymatela** [comment]
Currently in Hoff rebase&merges are always done in the order of PR approval.  So there should be no conflicts at all.

If B and C were approved in that order.  Then we would have `testing/B` rebased&merged on top of the new master and `testing/C` rebased&merged on top of `testing/B`.

This is already what happens in a scenario where a build fails on an early PR on the merge train:

![merge train restarts](https://github.com/channable/hoff/blob/master/doc/merge-train-restart.svg?sanitize=true)

Above, both PR#2 and PR#3 are due to be merged "at the same time".  But PR#3 is only merged slightly after PR#2 (and on top of it)

**Comment by rudymatela** [comment]
This scenario:

```
A->c---->m
--B->c---->c---->m
----C->c---->c----c--->m
```

happens more like this:

```
A->rmc---->p
--B->rmc-------->p
----C->rmc------------>p
```

where `rmc` is "rebase-then-merge-then-ci-job-started" and `p` is fastforwarding/pushing master to point to the new merge commit.

In the proposed solution, this is what would happen:

```
A->rmc-------rmc----->p
--B->rmc------rmc------->p
----C->rmc-----rmc-------->p
```

At the point we get a new commit to master, we "rebase-then-merge-then-ci-job-started" all pull requests in the order that they were approved.
